specifications/bcastByz/bcastByz.tla

```diff
@@ -0,0 +1,983 @@
+                           and Faulty are unchanged. *)
+VARIABLE pc             (* the control state of each process *)
+VARIABLE rcvd           (* the messages received by each process *)
+VARIABLE sent           (* the messages sent by all correct processes *)
+VARIABLE newMessages    (* the message received at the current step by each process *)
+
+ASSUME NTF == N \in Nat /\ T \in Nat /\ F \in Nat /\ (N > 3 * T) /\ (T >= F) /\ (F >= 0)
+
+Proc == 1 .. N                 (* all processess, including the faulty ones    *)
+ByzMsgs == { <<p, "ECHO">> : p \in Faulty }                            
+
+
+M == { "ECHO" }
+
+vars == << pc, rcvd, sent, newMessages, Corr, Faulty >>
+
+(* Both correct and faulty processes send only <ECHO> msgs. *)
+Receive(self) ==
+  /\ newMessages' \in SUBSET ( sent \cup ByzMsgs )
+  /\ rcvd' = [ i \in Proc |-> IF i # self THEN rcvd[i] ELSE rcvd[self] \cup newMessages' ]
+                             
+(* If received an INIT message and not send <ECHO> before,    
+ * then send <ECHO> to all - line 3.                                        
+ * UponV1 is for the 1st ITE.
+ *)
+UponV1(self) ==
+  /\ pc[self] = "V1"
+  /\ pc' = [pc EXCEPT ![self] = "SE"]
+  /\ sent' = sent \cup { <<self, "ECHO">> }
+  /\ UNCHANGED << Corr, Faulty >>
+
+(* If received <ECHO> from at least T + 1 distinct processes and   
+ * not sent <ECHO> before, then send <ECHO> to all. Since processes  
+ * send only <ECHO> msgs, #msgs = #distinct processes from which  
+ * process self received msgs. UponNonFaulty is for the 3rd ITE.
+ *)         
+UponNonFaulty(self) ==
+  /\ pc[self] /= "SE"
+  /\ Cardinality(rcvd'[self]) >= T + 1
+  /\ Cardinality(rcvd'[self]) < N - T
+  /\ pc' = [pc EXCEPT ![self] = "SE"]
+  /\ sent' = sent \cup { <<self, "ECHO">> }
+  /\ UNCHANGED << Corr, Faulty >>
+        
+(* If received <ECHO> from at least N - T distinct processes 
+ * and not send, <ECHO> before then accepts and sends <ECHO> to all.      
+ * UponAcceptNot is for the 2nd ITE when the 1st ITE was not satisfied
+ * and the 2nd ITE is satisfied.                 
+ *)        
+UponAcceptNotSent(self) ==
+  /\ pc[self] \in { "V0", "V1" }
+  /\ Cardinality(rcvd'[self]) >= N - T
+  /\ pc' = [pc EXCEPT ![self] = "AC"]
+  /\ sent' = sent \cup { <<self, "ECHO">> }
+  /\ UNCHANGED << Corr, Faulty >>
+        
+(* If received <ECHO> from at least N-T distinct processes before, accepts. 
+ * UponAccept is for the 2nd ITE when the 1st ITE was satisfied. 
+ *)        
+UponAccept(self) ==
+        /\ pc[self] = "SE"
+        /\ Cardinality(rcvd'[self]) >= N - T
+        /\ pc' = [pc EXCEPT ![self] = "AC"]
+        /\ sent' = sent
+        /\ UNCHANGED << Corr, Faulty >>
+
+(* Possible transition steps *)                
+Step(self) == 
+  /\ self \in Corr
+  /\ Receive(self)
+  /\ \/ UponV1(self)
+     \/ UponNonFaulty(self)
+     \/ UponAcceptNotSent(self)
+     \/ UponAccept(self)
+     \/ UNCHANGED << pc, sent, Corr, Faulty >>                 
```

Comment:
It's not quite clear what the UNCHANGED disjunct models: message reception without treatment by the receiver?